torsiondrive/extra_constraints.py

```diff
@@ -0,0 +1,198 @@
+#------------------------
+#  extra_constraints.py
+#------------------------
+from geometric.nifty import uncommadash, commadash
+
+def make_constraints_dict(constraints_string):
+    """ Create an ordered dictionary with constraints specification, consistent with geomeTRIC
+
+    Parameters
+    ----------
+    constraints_string: str
+        String-formatted constraint specification consistent with geomeTRIC constraints.txt
+
+    Returns
+    -------
+    constraints_dict: dict
+        A dictionary contains the definition of the extra constraints. The format is consistant with the JSON interface
+        of geomeTRIC. Example:
+        constraint_dict = {
+            'freeze': [{
+                "type": "xyz",
+                "indices": [0, 1, 2]
+            }],
+            'set': [{
+                "type": "angle",
+                "indices": [1, 0, 4],
+                "value": 110.0
+            }]
+        }
+
+    Notes
+    -----
+    1. Only Constraint of type "freeze" and "set" are supported, since extra "scan" is undefined with torsiondrive scan.
+    2. Five attributed are allowed to be constrained: 'bond','distance','angle','dihedral','xyz'
+    3. The input string is one-indexed, the output dictionary is zero-indexed.
+    4. For "xyz", dashed input like "1-3,7-9" (no space) is allowed, which will be converted to [0,1,2,6,7,8].
+    """
+    constraints_mode = None
+    constraints_dict = {'freeze': [], 'set': []}
+    for line in constraints_string.split('\n'):
+        # Ignore anything after a comment
+        line = line.split('#',1)[0].lower().strip()
+        if len(line) == 0: continue
+        if line.startswith('$'):
+            if line == '$freeze':
+                constraints_mode = 'freeze'
+            elif line == '$set':
+                constraints_mode = 'set'
+            elif line == '$end':
+                constraints_mode = None
+            elif line == '$scan':
+                raise ValueError('Additional $scan constraints are not allowed')
+            else:
+                raise ValueError('Unrecognized token starting with $')
+        else:
+            if constraints_mode == None:
+                raise ValueError(f'Trying to read the constraint line\n\n{line}\n\n, but constraint mode is not set')
```

Comment:
Oh cool, I hadn't seen this method of string formatting/variable substitution before.